src/goto-instrument/contracts/contracts.cpp

```diff
@@ -430,10 +444,41 @@ void code_contractst::check_apply_loop_contracts(
     }
 
     goto_function.body.destructive_insert(
-      loop_body, add_pragma_disable_assigns_check(generated_code));
+      loop_head, add_pragma_disable_assigns_check(generated_code));
   }
 
-  // Generate: assert(invariant) just after the loop exits
+  // Insert all the generated code so far:
+  // Assume invariant, decl the variant temporaries,
+  // and initialize old variant values.
+  goto_function.body.destructive_insert(
+    loop_head, add_pragma_disable_assigns_check(generated_code));
+
+  // Now we begin instrumenting at the loop_end.
+  // `generated_code` is to be inserted before `loop_end`.
+
+  // If we reach past the loop guard, then the loop runs at least once,
+  // so add the base case assertion:
+  //   assert(in_base_case ==> initial_invariant_value)
+  {
+    code_assertt assertion{
+      or_exprt{not_exprt{in_base_case}, initial_invariant_value}};
+    assertion.add_source_location() = loop_head->source_location();
+    converter.goto_convert(assertion, generated_code, mode);
+    generated_code.instructions.back().source_location_nonconst().set_comment(
+      "Check loop invariant before entry");
+  }
+
+  // and then jump back to the step case label
+  // to havoc the write set, assume the invariant,
+  // and execute an arbitrary iteration.
+  generated_code.add(goto_programt::make_goto(
+    step_case_target, in_base_case, loop_head->source_location()));
+
+  // The following code is only reachable in the step case,
+  // i.e., when in_base_case == false,
+  // because of the unconditional jump above.
+
+  // Generate: assert(invariant) just before the iteration ends
   // We use a block scope to create a temporary assertion,
```

Comment:
What does this mean ? why is the assertion called "temporary" ?

Reply:
So as not to pollute the function's namespace. The assertion is not temporary. It is immediate destroyed when we `goto_convert`, so there's no point keeping the destroyed object around anyway.